### PR TITLE
use auto if resolution is not specified

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-15554
build: https://builds.cask.co/browse/CDAP-DUT6986-1

Now, if a metrics query contains the time range but not the resolution, the minimum resolution is always used. This is incorrect since if the time range is > 10 hours, we are supposed to use the hour resolution, if > 10 mins, we are supposed to use min resolution. Using the second resolution will give many more data points.

### CODE REVIEW

File: cdap-gateway/src/test/java/io/cdap/cdap/gateway/handlers/metrics/MetricsHandlerTest.java [hunk 1]
```diff
@@ -780,6 +781,51 @@ public void testResultLimit() throws Exception {
...
+    verifyResolution("auto", 0L, 36000L, null, 200, "60s");
+    // if ts diff > 36000, hour resolution will be used
+    verifyResolution("auto", 0L, 36001L, null, 200, "3600s");
+
+    // if resolution is null, the resolution should be same if as resolution is auto if both timestamp are provided
+    verifyResolution(null, 0L, 300L, null, 200, "1s");
+    verifyResolution(null, 0L, 600L, null, 200, "1s");
+    verifyResolution(null, 0L, 601L, null, 200, "60s");
+    verifyResolution(null, 0L, 36000L, null, 200, "60s");
+    verifyResolution(null, 0L, 36001L, null, 200, "3600s");
```
Comment (anew) [inline cdap-gateway/src/test/java/io/cdap/cdap/gateway/handlers/metrics/MetricsHandlerTest.java:821]: this is a great test. It would be good to add cases where the start time is not 0L. Also, what happens if start > end? Should expect a 400?

File: cdap-watchdog/src/main/java/io/cdap/cdap/metrics/query/MetricsQueryHelper.java [hunk 2]
```diff
@@ -280,14 +280,33 @@ private void setTimeRangeInQueryRequest(MetricQueryRequest request, Map<String,
...
+   * 3. If the resolution is null, i.e, not specified in the query, then if start and end timestamp are not null,
+   * the logic will be same as the resolution is "auto". If any of the start or end timestamp is not specified,
+   * minimum resolution will be used.
+   *
+   * @param resolution the resolution string, can be specific resolution like 1s, 1m, etc, or can be auto or null.
+   * @param start the start timestamp, null if not specified in the query
+   * @param end the end timestamp, null if not specified in the query
+   * @return the integer resolution for the query
+   */
+  private Integer getResolution(@Nullable String resolution, @Nullable Long start, @Nullable Long end) {
```
Comment (anew) [inline cdap-watchdog/src/main/java/io/cdap/cdap/metrics/query/MetricsQueryHelper.java]: Actually, this method should have a unit test. The end2end test you have is great but it is better to test at the lowest level, that allows testing a lot more edge cases.

File: cdap-gateway/src/test/java/io/cdap/cdap/gateway/handlers/metrics/MetricsHandlerTest.java (hunk 1 above)
Comment (yaojiefeng) [reply cdap-gateway/src/test/java/io/cdap/cdap/gateway/handlers/metrics/MetricsHandlerTest.java:821]: 👍 will add more cases

File: cdap-watchdog/src/main/java/io/cdap/cdap/metrics/query/MetricsQueryHelper.java (hunk 2 above)
Comment (yaojiefeng) [reply cdap-watchdog/src/main/java/io/cdap/cdap/metrics/query/MetricsQueryHelper.java]: 👍 I wanted to add unit test for this class initially but essentially found there is not a single unit test for this class. And this method is just a small method in this class, I will add a unit test for this method now, and open a JIRA https://issues.cask.co/browse/CDAP-15558 to track other methods.

**Review by anew — APPROVED** [review]
